static/js/compose_ui.js

```diff
@@ -203,67 +196,51 @@ export function handle_keydown(event, textarea) {
+    }
 
     // detect Cmd and Ctrl key
     const isCmdOrCtrl = common.has_mac_keyboard() ? event.metaKey : event.ctrlKey;
 
-    if ((isBold || isItalic || isLink) && isCmdOrCtrl) {
-        const range = textarea.range();
-
-        if (isBold) {
-            // Ctrl + B: Convert selected text to bold text
-            wrap_text_with_markdown(textarea, "**", "**");
-            event.preventDefault();
-
-            if (!range.length) {
-                textarea.caret(textarea.caret() - 2);
-            }
-        }
-
-        if (isItalic) {
-            // Ctrl + I: Convert selected text to italic text
-            wrap_text_with_markdown(textarea, "*", "*");
-            event.preventDefault();
-
-            if (!range.length) {
-                textarea.caret(textarea.caret() - 1);
-            }
-        }
-
-        if (isLink) {
-            // Ctrl + L: Insert a link to selected text
-            wrap_text_with_markdown(textarea, "[", "](url)");
-            event.preventDefault();
-
-            const position = textarea.caret();
-            const txt = textarea[0];
-
-            // Include selected text in between [] parentheses and insert '(url)'
-            // where "url" should be automatically selected.
-            // Position of cursor depends on whether browser supports exec
-            // command or not. So set cursor position accordingly.
-            if (range.length > 0) {
-                if (document.queryCommandEnabled("insertText")) {
-                    txt.selectionStart = position - 4;
-                    txt.selectionEnd = position - 1;
-                } else {
-                    txt.selectionStart = position + range.length + 3;
-                    txt.selectionEnd = position + range.length + 6;
-                }
-            } else {
-                textarea.caret(textarea.caret() - 6);
-            }
-        }
-
+    if (type && isCmdOrCtrl) {
+        format_text(textarea, type);
         autosize_textarea(textarea);
-        return;
+        event.preventDefault();
     }
 }
 
 export function handle_keyup(event, textarea) {
     // Set the rtl class if the text has an rtl direction, remove it otherwise
     rtl.set_rtl_class_for_textarea(textarea);
 }
+
+export function format_text(textarea, type) {
+    const field = textarea.get(0);
+    const range = textarea.range();
+
+    switch (type) {
+        case "bold":
+            // Ctrl + B: Convert selected text to bold text
+            wrapSelection(field, "**");
+            break;
+        case "italic":
+            // Ctrl + I: Convert selected text to italic text
+            wrapSelection(field, "*");
+            break;
+        case "link":
+            // Ctrl + L: Insert a link to selected text
+            // Select `url`.
+            wrapSelection(field, "[", "](url)");
+
+            range.start = range.end + 3;
+            range.end = range.start + 3;
+            field.setSelectionRange(range.start, range.end);
```

Comment:
How does this logic work?  I don't super understand either the magic 3s or why we're swapping end/start.

Reply:
Added some comments.

Reply:
OK; I did some variable name renames so that it now reads:
```
+        case "link":
+            // Ctrl + L: Insert a link to selected text
+            wrapSelection(field, "[", "](url)");
+
+            // Change selected text to `url` part of the syntax.
+            // If <text> marks the selected region, we're mapping:
+            // <text> => [text](<url>).
+            const new_start = range.end + "[](".length;
+            const new_end = new_start + "url".length;
+            field.setSelectionRange(new_start, new_end);
+            break;
+    }
```

which clearly indicates where those 3s come from, and also avoid unnecessarily mutating the `range` variable.

Reply:
@timabbott Your edition adds an ESLint [`no-case-declaration`](https://eslint.org/docs/rules/no-case-declarations) error. You need braces around a case that includes variable declarations.